### PR TITLE
[linux-6.6.y] Driver for Zhaoxin SM3 SM4 algorithms

### PR DESCRIPTION
**Pull request by leoliu-oc**
Includes:
1. Add support for Zhaoxin GMI SM3 Secure Hash algorithm
This SM3 algorithm driver is developed to support the SM3 instruction,
making user develop their applications with both high performance and high
security.
2. Add support for Zhaoxin GMI SM4 Block Cipher algorithm
This SM4 algorithm driver is developed to support the SM4 instruction,
making user develop their applications with both high performance and high
security.

### CODE REVIEW

File: arch/x86/crypto/sm3_zhaoxin_gmi.c [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+		.cra_module      =  THIS_MODULE,
+	}
+};
+
+static int __init zx_sm3_generic_mod_init(void)
+{
+	if (gmi_available() == 0)
+		return crypto_register_shash(&zx_sm3_alg);
+
+	pr_warn("GMI is unavailable on this platform.");
```
Comment (opsiff) [inline arch/x86/crypto/sm3_zhaoxin_gmi.c]: 这里应该是pr_info 不应该是pr_warn吧 所有x86都会走

File: arch/x86/crypto/sm4_zhaoxin_gmi.c [hunk 2]
```diff
@@ -0,0 +1,859 @@
...
+			eax = 0xC0000001;
+			__asm__ __volatile__ ("cpuid":"=d"(edx):"a"(eax) : );
+
+			if ((edx & 0x0030) != 0x0030)
+				return -ENODEV;
+
+			pr_notice("GMI SM4 is detected by CPUID\n");
+			return 0;
+		}
+		pr_notice("GMI SM4 is available\n");
```
Comment (opsiff) [inline arch/x86/crypto/sm4_zhaoxin_gmi.c]: is available 为什么要用pr_notice?

File: arch/x86/crypto/sm4_zhaoxin_gmi.c [hunk 3]
```diff
@@ -0,0 +1,859 @@
...
+		|| (c->x86 > 6)) {
+		if (!boot_cpu_has(X86_FEATURE_CCS) || !boot_cpu_has(X86_FEATURE_CCS_EN)) {
+
+			eax = 0xC0000001;
+			__asm__ __volatile__ ("cpuid":"=d"(edx):"a"(eax) : );
+
+			if ((edx & 0x0030) != 0x0030)
+				return -ENODEV;
+
+			pr_notice("GMI SM4 is detected by CPUID\n");
```
Comment (opsiff) [inline arch/x86/crypto/sm4_zhaoxin_gmi.c]: is available 为什么要用pr_notice?

File: arch/x86/crypto/sm3_zhaoxin_gmi.c [hunk 4]
```diff
@@ -0,0 +1,203 @@
...
+		if (!boot_cpu_has(X86_FEATURE_CCS) ||
+			!boot_cpu_has(X86_FEATURE_CCS_EN)) {
+
+			eax = 0xC0000001;
+			__asm__ __volatile__ ("cpuid":"=d"(edx):"a"(eax) : );
+
+			if ((edx & 0x0030) != 0x0030)
+				return -ENODEV;
+
+			pr_notice("GMI SM3 detected by CPUID\n");
```
Comment (opsiff) [inline arch/x86/crypto/sm3_zhaoxin_gmi.c]: is available 为什么要用pr_notice?

File: arch/x86/crypto/sm3_zhaoxin_gmi.c [hunk 5]
```diff
@@ -0,0 +1,203 @@
...
+			return 0;
+		}
+		pr_notice("GMI SM3 is available\n");
```
Comment (opsiff) [inline arch/x86/crypto/sm3_zhaoxin_gmi.c]: is available 为什么要用pr_notice?

**Review by opsiff — CHANGES_REQUESTED** [review]

File: arch/x86/crypto/sm3_zhaoxin_gmi.c (hunk 1 above)
Comment (leoliu-oc) [reply arch/x86/crypto/sm3_zhaoxin_gmi.c]: 更新了一般，这里调整了逻辑。在匹配成功后才会pr_info打印。匹配不成功时不打印。这样就只有在兆芯平台才会有打印信息。

File: arch/x86/crypto/sm4_zhaoxin_gmi.c (hunk 2 above)
Comment (leoliu-oc) [reply arch/x86/crypto/sm4_zhaoxin_gmi.c]: 新版本中改为了pr_debug，供debug时判定用。正常情况下不输出log。

File: arch/x86/crypto/sm4_zhaoxin_gmi.c (hunk 3 above)
Comment (leoliu-oc) [reply arch/x86/crypto/sm4_zhaoxin_gmi.c]: 同上。

File: arch/x86/crypto/sm3_zhaoxin_gmi.c (hunk 4 above)
Comment (leoliu-oc) [reply arch/x86/crypto/sm3_zhaoxin_gmi.c]: 同上。

File: arch/x86/crypto/sm3_zhaoxin_gmi.c (hunk 5 above)
Comment (leoliu-oc) [reply arch/x86/crypto/sm3_zhaoxin_gmi.c]: 同上。

**Pull request merged**